### PR TITLE
Migrate to VexFlow 4

### PR DESCRIPTION
**Pull request by ronyeh**
This patch works with VexFlow 4.0.0-alpha-6, currently available at:

`npm install https://vexflow-demo.surge.sh/vexflow-4.0.0-alpha-6.tgz`

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -132,8 +132,8 @@ module.exports = grunt => {
                         'not ie <= 12',  // all versions -- edge is separate
                     ],
                 },
-                useBuiltIns: 'usage',
-                corejs: 3,
+                // useBuiltIns: 'usage',
+                // corejs: 3,
```
Comment (ronyeh) [inline Gruntfile.js:136]: Everything compiled file except babel/webpack were complaining about core-js not being resolved correctly by the vexflow library. VexFlow doesn't use core-js, and I couldn't find any direct imports of core-js in Music21j.

File: src/note.ts [hunk 2]
```diff
@@ -11,7 +11,7 @@
  *
  * @property {string[]} stemDirectionNames - an Array of allowable stemDirection names.
  */
-import Vex from 'vexflow';
+import { Accidental, StaveNote, Stem } from 'vexflow';
```
Comment (mscuthbert) [inline src/note.ts:14]: Because Music21 has an Accidental class as well, I will import this as VFAccidental.

File: src/vfShow.ts [hunk 3]
```diff
@@ -8,8 +8,23 @@
  * for rendering vexflow. Will eventually go to music21/converter/vexflow
  */
 
-import Vex from 'vexflow';
-
+import {
+    BarlineType,
+    Beam,
```
Comment (mscuthbert) [inline src/vfShow.ts:13]: This is a class that exists in music21 also.

File: src/vfShow.ts [hunk 4]
```diff
@@ -8,8 +8,23 @@
...
+    Formatter,
+    Fraction,
+    Renderer as VFRenderer,
+    Stave,
+    StaveConnector,
+    StaveNote,
+    StaveTie,
+    SVGContext,
+    TextNote,
+    Tuplet,
```
Comment (mscuthbert) [inline src/vfShow.ts:24]: This is a class that exists in music21

File: src/vfShow.ts [hunk 5]
```diff
@@ -8,8 +8,23 @@
...
+    Voice,
```
Comment (mscuthbert) [inline src/vfShow.ts:25]: Another class that also exists in music21  (I am sort of surprised this one wasn't already imported from music21...)

File: src/vfShow.ts [hunk 6]
```diff
@@ -1129,26 +1134,18 @@ export class Renderer {
...
         if (s === undefined) {
             s = this.stream;
         }
         const sClef = s.getSpecialContext('clef') || s.getContextByClass('Clef') || _clefSingleton;
         let noteOffsetLeft = 0;
         // var staveHeight = 80;
         if (stave !== undefined) {
-            noteOffsetLeft = stave.start_x + stave.glyph_start_x;
+            const noteStartX = stave.getNoteStartX();
+            noteOffsetLeft = noteStartX; // UNKNOWN: + stave.glyph_start_x;
```
Comment (mscuthbert) [inline src/vfShow.ts:1146]: This was an important fix at some point -- I can't remember which setting had the glyphs start at a different x than before, but I remember it being a lot of trial and error to find this one.  I might look at the "blame" to figure out when it came in.

File: Gruntfile.js (hunk 1 above)
Comment (mscuthbert) [reply Gruntfile.js:136]: see below -- it is used by music21j, and might be a bug in vexflow build if it doesn't work in this system.

Comment (mscuthbert) [reply Gruntfile.js:136]: and at least one system that uses music21j also uses core-js so if having Vexflow in the system means that core-js can't be used anywhere upstream then that would be a problem.

Comment (mscuthbert) [reply Gruntfile.js]: When I put core-js back in, it works perfectly for me.  Maybe I'm missing something in the dependencies list that I installed manually?  What's the error you are seeing?

**Comment by ronyeh** [comment]
Cool, if core-js works when you build, then it's probably fine. :-) I was just trying to get everything compiled with no errors and warnings. It was something about how some of the modern JS methods were unresolved (since VexFlow doesn't use core-js). If I can repro I'll send you an error log.

You don't need to treat this as an official PR. It's just a guide and you can apply the patch into a vex4 branch on your side.

I do agree you should use aliases for all imports of VexFlow classes. It's much clearer that way (VFOrnament vs Ornament).

**Comment by ronyeh** [comment]
If you plan to merge this into your master at some point (post VexFlow 4 release), let me know. I can update the PR with VFXXX prefixes for all the imports. That will clarify things and avoid name collisions.

**Comment by ronyeh** [comment]
I updated this PR with the most recent VexFlow 4 build. It only includes Gonville.

Give it a try and let me know if you run into any issues.

For me, the magic line seems to be this webpack resolve alias:
https://github.com/ronyeh/music21j/blob/103789c10ecbda86bc1c2aa8af98d1aed193c3a0/Gruntfile.js#L78-L80

https://webpack.js.org/configuration/resolve/#resolvealias

There are multiple ways to accomplish the same thing, apparently.

In my example projects, we use tsconfig.json's paths.

In the future, neither of these should be necessary, after TypeScript 4.6 and WebPack loaders work together to parse VexFlow's package.json properly. :-)

**Comment by mscuthbert** [comment]
Thanks -- I'll definitely get on this as soon as Vex4 is released fully!  :-)  good luck and thanks!

**Comment by mscuthbert** [comment]
Wonderful -- I will try this out sometime this weekend

File: package.json [hunk 7]
```diff
@@ -39,7 +39,6 @@
     "@typescript-eslint/parser": "^4.15.1",
     "ajv": "^7.1.1",
     "babel-loader": "^8.2.1",
-    "core-js": "^3.9.0",
```
Comment (mscuthbert) [inline package.json:42]: this I'll definitely need to put back.

**Comment by ronyeh** [comment]
Moved to here: https://github.com/cuthbertLab/music21j/pull/172

This PR was getting a bit messy due to repeated merges. The above PR is a bit cleaner.

**Pull request closed without merging**